### PR TITLE
Elasticity inversion using NEML2

### PR DESCRIPTION
**Pull request by dewenyushu**
Inverse optimization for material properties leveraging neml2 stress derivatives.

### CODE REVIEW

File: modules/tensor_mechanics/src/neml2/userobjects/CauchyStressFromNEML2UO.C [hunk 1]
```diff
@@ -45,11 +57,49 @@ CauchyStressFromNEML2UO::CauchyStressFromNEML2UO(const InputParameters & params)
...
+      mooseError("Batch material parameter is not ready for output");
+
+    auto model_param = model().named_parameters(true)[_reset_parameter_names[0]];
+    auto param_values = NEML2Utils::toNEML2Batched(_material_param_uo->getOutputData());
+
+    model_param = model_param.batch_expand_copy(param_values.batch_sizes());
+
+    model_param.copy_(param_values);
+  }
+}
```
Comment (dewenyushu) [inline modules/tensor_mechanics/src/neml2/userobjects/CauchyStressFromNEML2UO.C:101]: @hugary1995 Thanks for your recent updates in neml2 which makes the `batch_expand_copy` to expand the scalar to proper batch size without needing to hard-code it in the input file. 

However, I still find that the parameter is not correctly set and/or consumed in neml2 material model. Therefore, would you mind helping me check if this is the correct way of setting material properties? Or am I not calling `precompute()` in the proper location?

File: modules/combined/test/tests/material_inversion/test_param_set/model_test.i [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  [measure_data]
+    type = OptimizationData
+    variable = ux
+  []
+  [parametrization]
+    type = ConstantReporter
+    real_vector_names = 'coordx coordy youngs_modulus'
+    real_vector_values = '0 1 2; 0 1 2; 5 4 3'
+  []
+[]
```
Comment (dewenyushu) [inline modules/combined/test/tests/material_inversion/test_param_set/model_test.i:108]: @hugary1995 here is where I want to change the `youngs_modulus`. The idea is a bit complicated: 
 - set/transfer the `youngs_modulus` values at three physical locations in this reporter
 - extrapolate it in the function using `NearestReporterCoordinatesFunction`
 - transform the function to a material property `E_material` 
 - transform the material property as a batch material UO `E_batch_material`
 - then the UO is consumed in the `CauchyStressFromNEML2UO` for updating the neml2 material model parameter
 
So by changing the values here in the reporter, ideally we should see different displacement values. But I do not see the values changing.. I have checked that the values in `E_batch_material` changes correspondingly as I change the value in this reporter, which indicates an issue in setting the parameters in the `CauchyStressFromNEML2UO`

File: modules/tensor_mechanics/src/neml2/userobjects/CauchyStressFromNEML2UO.C (hunk 1 above)
Comment (dewenyushu) [reply modules/tensor_mechanics/src/neml2/userobjects/CauchyStressFromNEML2UO.C:101]: > Can this wait a few days as I wrap up the current (giant) NEML2 refactor?

No prob at all - thanks for considering this in your current neml2 refactor.

**Comment by dewenyushu** [comment]
I am confused about the documentation error and the phase_field & contact compilation error from civet. I cannot reproduce them on my local machine. Any ideas? Maybe @dschwen ?

**Comment by hugary1995** [comment]
Let me know if/when you want me to review.

File: modules/combined/test/tests/optimization/invOpt_elasticity/forward.i [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  [SolidMechanics]
+    [QuasiStatic]
+      [all]
+        strain = SMALL
+        new_system = true
+        add_variables = true
+        formulation = TOTAL
+        incremental = true
+        volumetric_locking_correction = false
+        # generate_output = 'cauchy_stress_xx'
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/forward.i:35]:
```suggestion
```

File: modules/combined/test/tests/optimization/invOpt_elasticity/forward.i [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    xmin = -4
+    xmax = 4
+    ymin = -4
+    ymax = 4
+  []
+[]
+
+[AuxVariables]
+  [dummy]
+  []
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/forward.i:20]: What's the use of this dummy variable?

File: modules/combined/test/tests/optimization/invOpt_elasticity/grad.i [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  [SolidMechanics]
+    [QuasiStatic]
+      [all]
+        strain = SMALL
+        new_system = true
+        add_variables = true
+        formulation = TOTAL
+        incremental = true
+        volumetric_locking_correction = false
+        # generate_output = 'cauchy_stress_xx'
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:57]:
```suggestion
```

File: modules/combined/test/tests/optimization/invOpt_elasticity/grad.i [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    base_name = 'forward'
+  []
+  # adjoint and forward use the same young's modulus value
+  [E_material]
+    type = GenericFunctionMaterial
+    prop_names = 'E_material'
+    prop_values = E
+  []
+[]
+[Functions]
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:112]:
```suggestion

[Functions]
```

File: modules/combined/test/tests/optimization/invOpt_elasticity/grad.i [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  [E]
+    type = NearestReporterCoordinatesFunction
+    x_coord_name = parametrization/coordx
+    y_coord_name = parametrization/coordy
+    value_name = parametrization/youngs_modulus
+  []
+[]
+[Reporters]
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:120]:
```suggestion

[Reporters]
```

File: modules/combined/test/tests/optimization/invOpt_elasticity/grad.i [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+[Executioner]
+  type = Steady
+  solve_type = NEWTON
+  petsc_options_iname = '-pc_type'
+  petsc_options_value = 'lu'
+  # we do not compute them together as this is overwritting DiracKernel's residual calculation, which should be improved
+  residual_and_jacobian_together = false
+[]
+[Outputs]
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:184]:
```suggestion

[Outputs]
```

File: modules/combined/test/tests/optimization/invOpt_elasticity/grad.i [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    function = E
+  []
+[]
+
+[Executioner]
+  type = Steady
+  solve_type = NEWTON
+  petsc_options_iname = '-pc_type'
+  petsc_options_value = 'lu'
+  # we do not compute them together as this is overwritting DiracKernel's residual calculation, which should be improved
```
Comment (hugary1995) [inline modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:181]: Could you open a ticket about this issue? We could use the new `mooseDocumentedError` API on this.

File: modules/optimization/doc/content/source/vectorpostprocessors/AdjointStrainBatchStressGradInnerProduct.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# AdjointStrainBatchStressGradInnerProduct
+
+This component is designed to compute the gradient of the objective function concerning specific properties. It achieves this by computing the inner product of the property derivative and the strain resulting from the forward simulation. The property derivative is extracted from a batch material of the [BatchPropertyDerivative](BatchPropertyDerivative.md) type.
```
Comment (hugary1995) [inline modules/optimization/doc/content/source/vectorpostprocessors/AdjointStrainBatchStressGradInnerProduct.md:3]: @lynnmunday If this is all you need, we could provide this 100 times more efficiently... We should probably talk about this.

NEML2 can give you this information in two ways:
1. Compute derivative of stress w.r.t. parameter, then compute its action on the adjoint strain.
2. Directly compute the action of the parameter derivative on the adjoint strain.

We are doing 1 right now. 2 is more efficient.

**Review by hugary1995 — APPROVED** [review]
Very exciting new feature! I just a few minor suggestions and a few things to discuss.

File: modules/combined/test/tests/optimization/invOpt_elasticity/forward.i (hunk 4 above)
Comment (dewenyushu) [reply modules/combined/test/tests/optimization/invOpt_elasticity/forward.i:20]: Good catch. It is actually not used in here.

File: modules/combined/test/tests/optimization/invOpt_elasticity/grad.i (hunk 9 above)
Comment (dewenyushu) [reply modules/combined/test/tests/optimization/invOpt_elasticity/grad.i:182]: Ref issue #27258

File: framework/include/userobjects/BatchPropertyDerivative.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  BatchPropertyDerivative(const InputParameters & params);
+
+  void batchCompute() override {}
+};
+
+typedef BatchPropertyDerivative<RankTwoTensor, Real> BatchPropertyDerivativeRankTwoTensorReal;
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply framework/include/userobjects/BatchPropertyDerivative.h:34]: Uhh, idk, maybe this is fine. I guess we are going to wrap everything inside an Action in the end, so this isn't that big of a concern. I am marking this as resolved now.

File: modules/optimization/doc/content/source/vectorpostprocessors/AdjointStrainBatchStressGradInnerProduct.md (hunk 10 above)
Comment (hugary1995) [reply modules/optimization/doc/content/source/vectorpostprocessors/AdjointStrainBatchStressGradInnerProduct.md:3]: Oh to be clear, we don't need to address this in this PR.

Comment (dewenyushu) [reply modules/optimization/doc/content/source/vectorpostprocessors/AdjointStrainBatchStressGradInnerProduct.md:3]: Noted in issue #26920

**Comment by dschwen** [comment]
This looks fine to me. I'm thinking about changes to batch material to allow runtime selection of gathered properties/variables etc. I'll have to do some profiling to see if all that template meta-programming hell was worth the effort, or if I can throw some of it out in favor of runtime code. But that shouldn't hold up this PR

**Review by dschwen — APPROVED** [review]

**Pull request merged**